File: 23-MergekSortedLists/23-MergekSortedLists.cpp
// Last updated: 7/13/2025, 10:26:43 PM
class Solution {
public:
    struct cmp {
        bool operator()(ListNode* a, ListNode* b) const {
            return a->val > b->val;           // min-heap on val
        }
    };

    ListNode* mergeKLists(vector<ListNode*>& v) {
        using Node = ListNode*;
        priority_queue<Node, vector<Node>, cmp> pq;

        /* push only the head of each non-empty list */
        for (Node head : v)
            if (head) pq.push(head);

        if (pq.empty()) return nullptr;       // every list was empty

        Node head = pq.top();                 // first (smallest) node
        Node tail = head;

        while (!pq.empty()) {
            tail = pq.top(); pq.pop();        // take smallest
            if (tail->next) pq.push(tail->next); // push successor
            if (!pq.empty()) tail->next = pq.top(); // link to new smallest
        }
        tail->next = nullptr;                 // terminate
        return head;
    }
};

File: 23-MergekSortedLists/23-MergekSortedLists.py
# Last updated: 3/14/2026, 6:06:53 PM
import heapq
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        values = []
        for node in lists:
            while node:
                values.append(node.val)
                node = node.next
        if not values:
            return None
        heapq.heapify(values)
        dummy = ListNode(0)
        curr = dummy
        while values:
            val = heapq.heappop(values)
            curr.next = ListNode(val)
            curr = curr.next
        return dummy.next



